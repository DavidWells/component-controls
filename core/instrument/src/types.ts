import { ParserOptions } from '@babel/parser';
import { SyncOpts as ResolveOptions } from 'resolve';
import {
  Options,
  ResolveConfigOptions as ResolvePrettierConfigOptions,
} from 'prettier';
import { PropTypes } from '@component-controls/specification';

type PrettierOptions = Options & {
  resolveConfigOptions?: ResolvePrettierConfigOptions;
};
export { ParserOptions, ResolveOptions, PrettierOptions };

export const defaultResolveOptions: ResolveOptions = {
  extensions: ['.js', '.jsx', '.ts', '.tsx', '.vue', '.mjs', '.es', '.es6'],
};

export const defaultParserOptions: ParserOptions = {
  sourceType: 'module',
  plugins: ['jsx', 'typescript'],
};

/**
 * callback function to extract props info table  - ie docgen type libraries
 * used to extract displayName, and props tables for a component
 */
export type PropsInfoExtractor = (
  /**
   * full name and path of the component path
   * react-docgen needs it to extract babel configurations
   */
  fileName: string,
  /**
   * optional component name
   * react-docgen-typescript supports multiple exports for a file
   * react-docgne does not use it
   */
  componentName?: string,
  /**
   * optional soure, saves time if its already loaded
   * react-docgen accepts source as input parameter
   * react-docgen-typescript does not use it
   */
  source?: string,
) => PropTypes | undefined;

export const defaultPackageOptions: PackageInfoOptions = {
  maxLevels: 10,
  packageJsonName: 'package.json',
  storeBrowseLink: true,
  storeDocsLink: true,
  storeIssuesLink: true,
};

export const defaultComponentOptions: ComponentOptions = {
  storeSourceFile: true,
  package: defaultPackageOptions,
};

export const defaultStoriesOptions: StoriesOptions = {
  storeSourceFile: true,
  package: defaultPackageOptions,
};

/**
 * options for finding and extracting package.json informtation
 * applies both for components and stories
 */
export interface PackageInfoOptions {
  /**
   * max levels of folders to look up to find the package.json file
   */
  maxLevels?: number;
  /**
   * package.json alternative name
   */
  packageJsonName?: string;

  /**
   * Whether to save the link for browsing the file in the repository field
   */
  storeBrowseLink?: boolean;

  /**
   * Whether to save the link for project readme file in the repository field
   */
  storeDocsLink?: boolean;

  /**
   * Whether to save the link for filing issues with the project in the repository field
   */

  storeIssuesLink?: boolean;
}

export interface ComponentOptions {
  /**
   * Callback function to resolve the source file name of a component.
   * Return false if this file should not be processed.
   */
  resolveFile?: (componentName: string, filePath: string) => string | undefined;

  /**
   * optional module to extract prop tables information for components
   */
  extractProps?: PropsInfoExtractor;
  /**
   * If set to false, will not save the component's source file
   */
  storeSourceFile?: boolean;

  /**
   * options for extracting repository information from the component's package,json file
   */
  package?: PackageInfoOptions | false;
}

export interface StoriesOptions {
  /**
   * If set to false, will not save the stories's source file, only the source of each individual story
   */
  storeSourceFile?: boolean;

  /**
   * options for extracting repository information from the component's package,json file
   */
  package?: PackageInfoOptions | false;
}

/**
 * Options passed for instrumenting
 */
export interface InstrumentOptions {
  /**
   * Options to control babel parser.
   */
  parser?: ParserOptions;
  /**
   * prettier options are used to prettify the code at the end of the process
   * this is useful for "story" code, where the story is extracted from the full file
   * fassing a value of false as prettier option will disabled prettifying
   */
  prettier?: PrettierOptions | false;

  /**
   * Options to control the `resolve` import resolving utilities.
   */
  resolver?: ResolveOptions;

  /**
   * Options for extracting components information.
   */
  components?: ComponentOptions;

  /**
   * Options for extracting stories information.
   */
  stories?: StoriesOptions;
}

/**
 * mdx-js options
 */
export interface MDXOptions {
  /**
   * enable footnotes
   */

  footnotes?: boolean;
  /**
   * specify remark plugins
   */
  mdPlugins?: any[];
  /**
   * specify rehype plugins
   */
  hastPlugins?: any[];
  /**
   * specify markdown compilers
   */
  compilers?: any[];

  /**
   * regex for blocks, defaults to ['[a-z\\.]+(\\.){0,1}[a-z\\.]']
   */
  blocks?: string[];
}

/**
 * instrument MDX has additional mdx options
 */
export interface InstrumentOptionsMDX extends InstrumentOptions {
  mdx?: MDXOptions;
}
