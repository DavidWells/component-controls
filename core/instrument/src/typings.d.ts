declare module 'read-package-json';
declare module '@hutson/parse-repository-url';
declare module 'hosted-git-info';
declare module '@mdx-js/mdx';
declare module 'js-string-escape';
declare module 'remark-emoji';
declare module 'remark-images';
