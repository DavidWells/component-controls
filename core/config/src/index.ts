import * as path from 'path';
import * as fs from 'fs';
import { sync as globSync } from 'glob';
import yargs from 'yargs';
import { Configuration } from '@component-controls/specification';

export const configFileNames = ['main.js', 'main.json', 'main.ts'];

export interface ConfigrationResult {
  config: Configuration;
  configPath: string;
}
/**
 * return the configration object
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the searh with
 */
export const getConfiguration = (
  baseFolder: string,
  args: string[] = process.argv,
): ConfigrationResult | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;

  const configPath = argv.config
    ? path.resolve(baseFolder, argv.config)
    : baseFolder;
  const configFiles = fs.readdirSync(configPath);
  const configFile = configFiles.find(file =>
    configFileNames.includes(file.toLowerCase()),
  );

  return configFile
    ? {
        config: require(path.resolve(configPath, configFile)),
        configPath,
      }
    : undefined;
};

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 * @param param0 configuration info return by getConfiguration
 */
export const extractStories = ({
  config,
  configPath,
}: ConfigrationResult): string[] | undefined => {
  const stories =
    config && config.stories
      ? config.stories.reduce((acc: string[], storyRg: string) => {
          const regex = storyRg.replace(/\((.*?)\)/g, '+($1)');
          return [...acc, ...globSync(path.resolve(configPath, regex))];
        }, [])
      : undefined;
  return stories;
};
