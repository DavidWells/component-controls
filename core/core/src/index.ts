export * from './utils';
export { randomizeData } from './randomizeData';
export * from './smartControls';
export * from './source';
export * from './faker';
