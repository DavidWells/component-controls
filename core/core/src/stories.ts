import { CodeLocation, PackageInfo, StoryRenderFn } from './utility';
import { StoryComponent } from './components';
import { ComponentControls } from './controls';
import { RunConfiguration, PagesOnlyRoutes, PageType } from './configuration';
/**
 * an identifier/variable.argument in the source code
 */
export interface SourceIdentifier {
  name: string;
  loc?: CodeLocation;
}
export interface ArgUsageLocation {
  /**
   * where in the story source code is the argument used
   * code location is relative to the start of the story
   */
  loc: CodeLocation;

  /**
   * optional name for the usage of the argument
   * example: export const story = ({ value }) => <Story value={{ age: value }} />;
   * in this example the name will be 'age'
   */
  name?: SourceIdentifier;

  /** true if the property is a 'shorthand'.
   * { prop: value } - not a shorthand.
   * { prop } - a shorthand.
   */
  shorthand?: boolean;
}

/**
 * arguments passed to the 'story' function, extracted by an AST loader
 */
export interface StoryArgument {
  /**
   * either the name used (or a variable alias), or an array of sub-arguments
   * ({ name: alias })
   */
  value: string | StoryArguments;
  /**
   * the original name of the argument
   */
  name?: string;
  /**
   * location of the argument declaration, relative to the story source code
   */
  loc?: CodeLocation;
  /**
   * list of locations where the argument is used in the body of the story
   */
  usage?: ArgUsageLocation[];
}

/**
 * list of story arguments. Each argument can be a deconstructed argument of itself
 * the first argument are the control 'values'
 */
export type StoryArguments = StoryArgument[];

/**
 * list of configuration parameters for indivudual stories and docs files
 * can be specified either through CSF or MDX tags
 */
export interface StoryParameters {
  [name: string]: any;
}

/**
 * Story interface - usually extracted by the AST instrumenting loader
 */
export interface Story {
  /**
   * name of the Story.
   */
  name: string;

  /**
   * csf id of the story
   */
  id?: string;

  /**
   * title of the file/group of stories
   */
  doc?: string;

  /**
   * render function for the story
   */
  renderFn?: StoryRenderFn;

  /**
   * story extended description. can use markdown.
   */
  description?: string;

  /**
   * arguments pass to a CSF story
   * eg `export const story = props => <Story {...props} />;`
   */
  arguments?: StoryArguments;
  /**
   * configuration parameters passed to the story - either CSF or MDX
   */
  parameters?: StoryParameters;
  /**
   * location in the source file of the story definition
   */
  loc?: CodeLocation;
  /**
   * the source code of the story, extracted byt the AST instrumenting loaders
   */
  source?: string;

  /**
   * optional story subtitle property
   */
  subtitle?: string;

  /**
   * id for component associated with the story
   */
  component?: string | object;

  /**
   * multiple components option
   */
  subcomponents?: {
    [key: string]: string | object;
  };

  /**
   * object of key/value pairs specifying the controls for the story
   */
  controls?: ComponentControls;

  /**
   * story decorators (or wrappers)
   */
  decorators?: StoryRenderFn[];
}

/**
 * map of stories. The id is compatible with CSF story ids
 */
export interface Stories {
  [id: string]: Story;
}

export const defPageType: PageType = 'story';
/**
 * A documentation file's metadata.
 * For MDX files, fromtmatter is used to declare the document properties.
 * For "es modules" documetation files, the default export is used.
 */
export interface Document {
  /**
   * title of the document. If no 'route' parameter is specifified, the title is used to generate the document url.
   * This is the only required field, to show the document in the menu structures.
   */
  title: string;

  /**
   * document type - blogs, pages, stories and even custom ones. By default - story
   */
  type?: PageType;

  /**
   * if provided, will be used as the route for the page.
   * if not provided, the title in lowercase will be used as the route
   */
  route?: string;

  /**
   * configuration parameters passed to the story groups
   * configured either as CSF default export
   * or MDX <Meta /> tag
   */
  parameters?: StoryParameters;

  /**
   * id for component associated with the stories file
   */
  component?: string | object;

  /**
   * multiple components option
   */
  subcomponents?: string[] | object[];

  /**
   * story decorators (or wrappers)
   */
  decorators?: StoryRenderFn[];

  /**
   * object of key/value pairs specifying the controls for the stories file
   * this will apply to all the stories in the file
   */
  controls?: ComponentControls;

  /**
   * if true, will display the documentation page full size (pagecontainer.full theme variant)
   * the default value is from the page type configuration
   */
  fullPage?: boolean;

  /**
   * whether to add navigation sidebars to the page
   * the default value is from the page type configuration
   */
  sidebars?: boolean;

  /**
   *  optional date the document was created. If not assigned, the instrumentation process will use birthtime
   */
  date?: Date;

  /**
   *  optional date the document was last modified. If not assigned, the instrumentation process will use mtime
   */
  dateModified?: Date;

  /**
   *  comma-separated list of document tags, used for search
   */
  tags?: string[];

  /**
   *  documentation file description
   */
  description?: string;

  /**
   * document author
   */
  author?: string;

  /**
   * document order, used to sort documents within the same parent
   */
  order?: number;

  //next generated by the instrumentation process

  /**
   * list of story ids contained in the document.
   */
  stories?: string[];

  /**
   * source code of the entire file of stories
   */
  source?: string;

  /**
   * file name of the file of stories
   */
  fileName?: string;

  /**
   * lookup into the global store of PackageInfo package.json
   */
  package?: string;

  /**
   * lookup into the global store.components
   * since multiple components of the same name can be used
   * example: ['Button']: 'c:/myapp/Button.tsx'
   */
  componentsLookup: {
    [name: string]: string;
  };

  /**
   * for MDX documents, this is an MDXContent function, to be rendered inside a MDXProvider
   */
  MDXPage?: any;
}

export const dateToLocalString = (date?: Date): string =>
  date
    ? new Date(date).toLocaleDateString('en-US', {
        timeZone: 'UTC',
      })
    : '';
/**
 * list of components used in stories
 */

export interface StoryComponents {
  [fileName: string]: StoryComponent;
}

/**
 * list of story files, or groups
 */
export interface Documents {
  [title: string]: Document;
}

export type Pages = Document[];

/**
 * list of stories
 */
export interface StoryStories {
  [id: string]: Story;
}

/**
 * list of repositories
 */
export interface StoryPackages {
  [id: string]: PackageInfo;
}

/**
 * store of stories information in memory after the loader is applied
 */
export interface StoriesStore {
  /**
   * global configuration for config file
   */
  config?: RunConfiguration;
  /**
   * list of story files, or groups
   */
  docs: Documents;
  /**
   * list of stories
   */
  stories: StoryStories;
  /**
   * list of components used in stories
   */
  components: StoryComponents;

  /**
   * list of package.json files and their data
   * used by the components and the stories of the project
   */
  packages: StoryPackages;
}

export const getDocPath = (
  pageType: PageType,
  doc?: Document,
  pagesConfig?: PagesOnlyRoutes,
  name: string = '',
  activeTab?: string,
): string => {
  const { basePath = '' } = pagesConfig?.[pageType] || {};
  const route = doc
    ? doc.route ||
      `/${basePath}${
        activeTab ? `${activeTab}/` : ''
      }${doc.title?.toLowerCase()}/`
    : `/${basePath}${activeTab ? `${activeTab}/` : ''}${name}/`;
  return route;
};

export const getStoryPath = (
  story?: Story,
  doc?: Document,
  pagesConfig?: PagesOnlyRoutes,
  activeTab?: string,
): string => {
  if (!story) {
    return '';
  }

  const docsPath = getDocPath('story', doc, pagesConfig, undefined, activeTab);
  return `${docsPath}#${story.id}`;
};
