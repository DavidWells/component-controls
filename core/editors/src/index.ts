export * from './editors/prop-factory';
export * from './editors/types';
export * from './forms/ControlsTable/ControlEditorsTable';
export { FlexContainer } from './components/FlexContainer/FlexContainer';
