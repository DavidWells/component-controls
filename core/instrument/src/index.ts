import * as parser from '@babel/parser';
const mdx = require('@mdx-js/mdx');
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import prettier, { Options, ResolveConfigOptions } from 'prettier';
import parserBabel from 'prettier/parser-babylon';
import { StoriesGroup, Story } from '@component-controls/specification';
import { extractCSFStories } from './babel/csf-stories';
import { extractMDXStories } from './babel/mdx-stories';
import { removeMDXAttributes } from './babel/remove-mdx-attributes';

type TraverseFn = (stories: StoriesGroup) => any;

export type PrettierOptions = Options & {
  resolveConfigOptions?: ResolveConfigOptions;
};
const parseSource = async (
  code: string,
  traverseFn: TraverseFn,
  originalSource: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  let source: string;
  if (prettierOptions !== false) {
    const { resolveConfigOptions, ...otherOptions } = prettierOptions || {};
    let allPrettierOptions = otherOptions;
    if (filePath) {
      const userOptions = await prettier.resolveConfig(
        filePath,
        resolveConfigOptions,
      );
      allPrettierOptions = { ...userOptions, ...allPrettierOptions };
    }

    source = prettier.format(code, {
      parser: 'typescript',
      plugins: [parserBabel],
      ...allPrettierOptions,
    });
  } else {
    source = code;
  }

  const ast = parser.parse(source, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript'],
  });
  const stories = {
    stories: {},
    source: originalSource,
  };
  traverse(ast, traverseFn(stories));
  Object.keys(stories.stories).forEach((key: string) => {
    //@ts-ignore
    const story: Story = stories.stories[key];
    const { start, end } = story.loc || {};
    if (start && end) {
      const lines = source.split('\n');

      if (start.line === end.line) {
        story.source = lines[start.line - 1].substring(
          start.column,
          end.column,
        );
      } else {
        const startLine = lines[start.line - 1];
        const endLine = lines[end.line - 1];
        if (startLine !== undefined && endLine !== undefined) {
          story.source = [
            startLine.substring(start.column),
            ...lines.slice(start.line, end.line - 1),
            endLine.substring(0, end.column),
          ].join('\n');
        }
      }
    }
  });
  return stories;
};

export const parseCSF = async (
  source: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  return await parseSource(
    source,
    extractCSFStories,
    source,
    filePath,
    prettierOptions,
  );
};

export const parseMDX = async (
  source: string,
  filePath?: string,
  prettierOptions?: PrettierOptions,
): Promise<StoriesGroup> => {
  const code = await mdx(source);

  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript'],
  });
  traverse(ast, removeMDXAttributes());
  const newCode = generate(ast, {
    retainFunctionParens: true,
    retainLines: true,
  });
  return await parseSource(
    newCode.code,
    extractMDXStories,
    source,
    filePath,
    prettierOptions,
  );
};
