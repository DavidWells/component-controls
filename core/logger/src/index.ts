import chalk from 'chalk';

export type LogLevel = 'all' | 'errors' | 'none';

export interface LogOptions {
  logLevel?: LogLevel;
  colors: [number, number, number];
}

const defaultOptions: LogOptions = {
  logLevel: 'all',
  colors: [244, 147, 66],
};

let logOptions: LogOptions = { ...defaultOptions };

const output = (heading: string, text: string) => {
  console.log(chalk.bgRgb(...logOptions.colors)(`@${heading}`), text);
};

export const log = (heading: string, text: string): void => {
  if (logOptions.logLevel === 'all') {
    output(heading, text);
  }
};
export const error = (heading: string, text: string): void => {
  if (logOptions.logLevel !== 'none') {
    output(heading, text);
  }
};

export const setLogOptions = (options: Partial<LogOptions> = {}): void => {
  logOptions = { ...logOptions, ...options };
};
