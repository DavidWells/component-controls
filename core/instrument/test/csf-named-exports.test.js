import { parseCSF } from '../src/index';

describe('csf-named-exports', () => {
  it('story propery name', async () => {
    expect(
      await parseCSF(`
const myStory = () => {};
myStory.story = {
 name: 'Custom story name',
}

export { myStory }
    `),
    ).toMatchSnapshot();
  });

  it('exported alias name', async () => {
    expect(
      await parseCSF(`
const myStory = () => {};
myStory.story = {
 name: 'Custom story name',
}

export { myStory as exportedStory}
    `),
    ).toMatchSnapshot();
  });

  it('re-exported name', async () => {
    expect(
      await parseCSF(`
import { myStory } from 'stories.tsx';
myStory.story = {
  name: 'Custom story name',
};

export { myStory };
    
    `),
    ).toMatchSnapshot();
  });
});
