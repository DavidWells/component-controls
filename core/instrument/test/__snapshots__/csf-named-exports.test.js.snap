// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csf-named-exports exported alias name 1`] = `
Object {
  "source": "
const myStory = () => {};
myStory.story = {
 name: 'Custom story name',
}

export { myStory as exportedStory}
    ",
  "stories": Object {
    "exportedStory": Object {
      "arguments": Array [],
      "loc": Object {
        "end": Object {
          "column": 24,
          "line": 1,
        },
        "start": Object {
          "column": 16,
          "line": 1,
        },
      },
      "name": "Custom story name",
      "parameters": Array [
        Object {
          "loc": Object {
            "end": Object {
              "column": 27,
              "line": 3,
            },
            "start": Object {
              "column": 2,
              "line": 3,
            },
          },
          "name": "name",
          "value": "Custom story name",
        },
      ],
      "source": "() => {}",
    },
  },
}
`;

exports[`csf-named-exports re-exported name 1`] = `
Object {
  "source": "
import { myStory } from 'stories.tsx';
myStory.story = {
  name: 'Custom story name',
};

export { myStory };
    
    ",
  "stories": Object {},
}
`;

exports[`csf-named-exports story propery name 1`] = `
Object {
  "source": "
const myStory = () => {};
myStory.story = {
 name: 'Custom story name',
}

export { myStory }
    ",
  "stories": Object {
    "myStory": Object {
      "arguments": Array [],
      "loc": Object {
        "end": Object {
          "column": 24,
          "line": 1,
        },
        "start": Object {
          "column": 16,
          "line": 1,
        },
      },
      "name": "Custom story name",
      "parameters": Array [
        Object {
          "loc": Object {
            "end": Object {
              "column": 27,
              "line": 3,
            },
            "start": Object {
              "column": 2,
              "line": 3,
            },
          },
          "name": "name",
          "value": "Custom story name",
        },
      ],
      "source": "() => {}",
    },
  },
}
`;
